### PR TITLE
feat: Improve error experience

### LINKED ISSUE #2842: Improve errors user experience
Fastify has many assertions around in the codebase to make sure that the user didn't misconfigure something, for instance, it checks if a route has been declared twice or if a plugin is not a function.
This is great, we vastly reduce the possibility of weird runtime issues, but for historical reasons the errors we are emitting are everything but user friendly.

We should improve this situation, as it is a big pain point in our user experience.

Example of double route declaration:
```js
'use strict'
const fastify = require('fastify')()
fastify.get('/', () => {})
fastify.get('/', () => {})
fastify.listen(3000, console.log)
```

```
▶ node index.js
AssertionError [ERR_ASSERTION]: Method 'GET' already declared for route '/'
    at Router._insert (.../fst-err/node_modules/find-my-way/index.js:304:9)
    at Router._on (.../fst-err/node_modules/find-my-way/index.js:197:8)
    at Router.on (.../fst-err/node_modules/find-my-way/index.js:74:8)
    at Object.afterRouteAdded (.../fst-err/node_modules/fastify/lib/route.js:253:16)
    at .../fst-err/node_modules/fastify/lib/route.js:187:25
    at Object._encapsulateThreeParam (.../fst-err/node_modules/avvio/boot.js:545:7)
    at Boot.timeoutCall (.../fst-err/node_modules/avvio/boot.js:447:5)
    at Boot.callWithCbOrNextTick (.../fst-err/node_modules/avvio/boot.js:428:19)
    at Boot._after (.../fst-err/node_modules/avvio/boot.js:273:26)
    at Plugin.exec (.../fst-err/node_modules/avvio/plugin.js:131:17) {
  generatedMessage: false,
  code: 'ERR_ASSERTION',
  actual: false,
  expected: true,
  operator: '=='
}
```

What I would expect instead is:
```
FastifyError: Method 'GET' already declared for route '/'
   at fastify.get('/', () => {}) (.../fst-err/index.js:4:9)
```
Which is quite complex, given that our route declaration is asynchronous.

Another example, this time with bad plugin type:
```js
'use strict'
const fastify = require('fastify')()
fastify.register(null)
fastify.listen(3000, console.log)
```
```
▶ node index.js
.../fst-err/node_modules/avvio/boot.js:200
    throw new Error('plugin must be a function or a promise')
    ^

Error: plugin must be a function or a promise
    at assertPlugin (.../fst-err/node_modules/avvio/boot.js:200:11)
    at Boot._addPlugin (.../fst-err/node_modules/avvio/boot.js:233:12)
    at Boot.use (.../fst-err/node_modules/avvio/boot.js:209:25)
    at Object.server.<computed> [as register] (.../fst-err/node_modules/avvio/boot.js:33:14)
    at Object.<anonymous> (.../fst-err/index.js:5:9)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:72:12)
```

As you can see, the error is not being helpful at all.

### PR DESCRIPTION
**Pull request by MetCoder95**
Fixes #2842 

This PR tries to improve the error experience by changing to sync registering of routes along with a new error object for Fastify errors.

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: fastify.js [hunk 1]
```diff
@@ -502,6 +502,9 @@ function fastify (options) {
     } else if (name === 'onReady') {
       this[kHooks].validate(name, fn)
       this[kHooks].add(name, fn)
+    } else if (name === 'onRoute') {
```
Comment (MetCoder95) [inline fastify.js:505]: I need to move the setup for the `onRoute` hooks due to the fact that was made for being async, in order to achieve register routes sync, the hooks should also be set in the same way. Please let me know your thoughts

File: lib/route.js [hunk 2]
```diff
@@ -154,43 +155,61 @@ function buildRouting (options) {
...
       }
-    })
+    } else if (path[0] === '/' && prefix.endsWith('/')) {
+      // Ensure that '/prefix/' + '/route' gets registered as '/prefix/route'
+      addNewRoute.call(this, path.slice(1))
+    } else {
+      addNewRoute.call(this, path)
+    }
+
+    // TODO: remove once review is done
```
Comment (MetCoder95) [inline lib/route.js:182]: I'm just preserving these comments as a base for doing further changes if needed based on the old implementation. Once everything approved, I'll delete them

File: test/hooks.test.js [hunk 3]
```diff
@@ -740,24 +740,25 @@ test('onRoute hook should able to change the route url', t => {
...
-      reply.send()
-    })
-    done()
-  })
-
-  fastify.ready(err => {
-    t.ok(err)
-  })
-})
+// TODO: remove once review is done
```
Comment (MetCoder95) [inline test/hooks.test.js:743]: Now that also the `onRoute` hooks are sync, the error handling needs to also be sync. Just want to double-check and get feedback before modifying the test

File: lib/route.js [hunk 4]
```diff
@@ -154,43 +155,61 @@ function buildRouting (options) {
...
 
-    /**
-     * This function sets up a new route, its log serializers, and triggers route hooks.
-     *
-     * @param {object} opts contains route `path` and `prefixing` flag which indicates if this is an auto-prefixed route, e.g. `fastify.register(routes, { prefix: '/foo' })`
-     * @param {*} notHandledErr error object to be passed back to the original invoker
-     * @param {*} done callback
-     */
-    function afterRouteAdded ({ path, prefixing = false }, notHandledErr, done) {
+    function addNewRoute (path, prefixing = false) {
```
Comment (mcollina) [inline lib/route.js:212]: I would try to move this to an arrow function to avoid calling it with `this`.

File: lib/route.js [hunk 5]
```diff
@@ -266,6 +277,103 @@ function buildRouting (options) {
...
+    }
+
+    /**
+     * This function sets up a new route, its log serializers, and triggers route hooks.
+     *
+     * @param {object} opts contains route `path` and `prefixing` flag which indicates if this is an auto-prefixed route, e.g. `fastify.register(routes, { prefix: '/foo' })`
+     * @param {*} notHandledErr error object to be passed back to the original invoker
+     * @param {*} done callback
+     */
+    function afterRouteAdded (path, url, context, notHandledErr, done) {
```
Comment (mcollina) [inline lib/route.js:302]: please embed the content of this function inside the previous `.after()` call.

File: test/hooks.test.js (hunk 3 above)
Comment (mcollina) [reply test/hooks.test.js:743]: It's totally ok to handle the error synchronously

**Review by mcollina — CHANGES_REQUESTED** [review]
Can you please target `next` instead?

**Comment by jsumners** [comment]
The description as I write this is "WIP". Therefore, I am marking it as a draft. Please mark it ready for review when you feel it is ready.

File: lib/route.js (hunk 4 above)
Comment (MetCoder95) [reply lib/route.js:212]: Sure, but that would make the function unreachable and will need to either declare it outside of `buildRouting` or in someplace within [route.js:L43-L45](https://github.com/fastify/fastify/blob/next/lib/route.js#L43-L45)

Or there's some other way to do it and I'm missing it?

File: lib/route.js [hunk 6]
```diff
@@ -208,51 +228,57 @@ function buildRouting (options) {
...
+      // adding a new route
+
+      // There's a bug when ignoreTrailingSlash !== true
+      // See
+      /*
+      └── /prefix
+          └── / (GET)
+              / (HEAD)
+              └── / (GET)
+      **/
```
Comment (MetCoder95) [inline lib/route.js:269]: This is a pretty interesting error popping up only when `ignoreTrailingSlash === true`, I didn't dig deep too much but was quite hard to detect. The third route comes out nowhere and the call stack is not really descriptive. I can take a look into it but not sure if it belongs directly to this PR as it happens before targeting to `next` branch.

File: test/route.test.js [hunk 7]
```diff
@@ -173,7 +173,7 @@ test('route', t => {
 test('invalid schema - route', t => {
   t.plan(3)
 
-  const fastify = Fastify()
+  const fastify = Fastify({ exposeHeadRoutes: false })
```
Comment (MetCoder95) [inline test/route.test.js:176]: I set `exposeHeadRoutes` to false in order to let the test pass with the correct route, as now if there's a schema validation error, the `HEAD` route validation is being thrown first, just before the origin `GET` route

File: test/404s.test.js [hunk 8]
```diff
@@ -830,21 +830,19 @@ test('setNotFoundHandler should not suppress duplicated routes checking', t => {
...
+  try {
+    fastify.get('/', function (req, reply) {
+      reply.send({ hello: 'world' })
+    })
+    fastify.get('/', function (req, reply) {
+      reply.send({ hello: 'world' })
+    })
+    fastify.setNotFoundHandler(function (req, reply) {
+      reply.code(404).send('this was not found')
+    })
```
Comment (mcollina) [inline test/404s.test.js:842]: please add a t.fail() at the end of the try block.

File: test/route.test.js (hunk 7 above)
Comment (mcollina) [reply test/route.test.js:176]: Have you rebased this PR on top of `next`? I think I fixed this.

File: test/schema-feature.test.js [hunk 9]
```diff
@@ -170,7 +170,7 @@ test('Cannot add schema for query and querystring', t => {
 
 test('Should throw of the schema does not exists in input', t => {
   t.plan(2)
-  const fastify = Fastify()
+  const fastify = Fastify({ exposeHeadRoutes: false })
```
Comment (mcollina) [inline test/schema-feature.test.js:173]: what happens if this is set to true?

File: test/throw.test.js [hunk 10]
```diff
@@ -17,12 +17,13 @@ test('Fastify should throw on wrong options', t => {
...
   t.plan(1)
   const fastify = Fastify()
-  fastify.get('/', () => {})
-  fastify.get('/', () => {})
 
-  fastify.ready(err => {
-    t.is(err.message, "Method 'GET' already declared for route '/' with constraints '{}'")
-  })
+  try {
+    fastify.get('/', () => {})
```
Comment (mcollina) [inline test/throw.test.js:22]: please move the first `get()` outside of the try-catch.

File: lib/route.js (hunk 6 above)
Comment (mcollina) [reply lib/route.js:269]: I don't understand the error you are describing. As it happens on `master` as well, open a fresh issue.

**Review by mcollina — CHANGES_REQUESTED** [review]

File: test/hooks.test.js [hunk 11]
```diff
@@ -740,23 +740,27 @@ test('onRoute hook should able to change the route url', t => {
   })
 })
 
-test('onRoute hook that throws should be caught', t => {
+test('onRoute hook that throws should be caught', { only: true }, t => {
```
Comment (jsumners) [inline test/hooks.test.js:743]:
```suggestion
test('onRoute hook that throws should be caught', t => {
```

File: test/throw.test.js [hunk 12]
```diff
@@ -17,12 +17,13 @@ test('Fastify should throw on wrong options', t => {
...
+    fastify.get('/', () => {})
```
Comment (RafaelGSS) [inline test/throw.test.js:23]: I'd add a t.fail bellow of this statement.

File: test/hooks.test.js [hunk 13]
```diff
@@ -740,23 +740,27 @@ test('onRoute hook should able to change the route url', t => {
...
       throw new Error('snap')
     })
-    instance.get('/', opts, function (req, reply) {
-      reply.send()
-    })
+
+    try {
+      instance.get('/', opts, function (req, reply) {
+        reply.send()
+      })
```
Comment (RafaelGSS) [inline test/hooks.test.js:755]: Same here. `t.fail` with a declarative description.

File: lib/route.js (hunk 6 above)
Comment (MetCoder95) [reply lib/route.js:269]: Not sure if it happens on `main` but caused me problems when pointing to `next`.
What it does is create a third route out of nowhere, as described in the comment. A third route under `/prefix//` is made, only if `ignoreTrailingSlash = true`. I'll try to see if I can reproduce it on `main`

File: test/schema-feature.test.js (hunk 9 above)
Comment (MetCoder95) [reply test/schema-feature.test.js:173]: The error validates a throw on an unexisting schema reference, but the test is waiting for the error to be thrown on the original `GET /:id` endpoint. But because of the order of the routes being loaded (first `GET` , second `HEAD`), the validation always fails on the `HEAD` route throwing exactly the same error but just for the sibling route

File: test/route.test.js (hunk 7 above)
Comment (MetCoder95) [reply test/route.test.js:176]: Yes, actually rebasing to `next` popped up the issue. The routes are registered in order, first `GET` and `HEAD` as second, but as the schema build happens async, the first that comes up is always for the `HEAD` one

File: test/hooks.test.js (hunk 13 above)
Comment (MetCoder95) [reply test/hooks.test.js:755]: Done in 0849c1b 👍

File: test/throw.test.js (hunk 12 above)
Comment (MetCoder95) [reply test/throw.test.js]: Done in 0849c1b 👍

File: test/throw.test.js (hunk 10 above)
Comment (MetCoder95) [reply test/throw.test.js:24]: Done in 0849c1b 👍

File: test/hooks.test.js (hunk 11 above)
Comment (MetCoder95) [reply test/hooks.test.js]: Done in 0849c1b 👍

File: test/schema-feature.test.js (hunk 9 above)
Comment (mcollina) [reply test/schema-feature.test.js:173]: Can you add a test with `exposeHeadRoutes: true`?

File: test/route.test.js (hunk 7 above)
Comment (mcollina) [reply test/route.test.js:176]: I have a feeling that we should fix this.

Comment (MetCoder95) [reply test/route.test.js]: I found the issue, basically in order to receive first the error thrown by the schema validation of the first `GET` route, is a need to pass the error to the `done` function when receiving the error from the schema validation. Solved on 1ac09cc

**Comment by MetCoder95** [comment]
There was a recent update that changed the behavior of `fast-json-stringify`?

The tests is receiving a new error:

```text
            -Failed building the serialization schema for GET: /:id, due to error Cannot read property 'type' of undefined
          
            +Failed building the serialization schema for GET: /:id, due to error schema is invalid: data.properties['name'] should be object,boolean
```

File: lib/route.js [hunk 14]
```diff
@@ -208,117 +228,132 @@ function buildRouting (options) {
...
+
+        const { exposeHeadRoute } = opts
+        const hasRouteExposeHeadRouteFlag = exposeHeadRoute != null
+        const shouldExposeHead = hasRouteExposeHeadRouteFlag ? exposeHeadRoute : globalExposeHeadRoutes
+
+        if (shouldExposeHead && options.method === 'GET' && !headRouteExists) {
+          const onSendHandlers = parseHeadOnSendHandlers(opts.onSend)
+          prepareRoute.call(this, 'HEAD', path, { ...opts, onSend: onSendHandlers })
+        } else if (headRouteExists && exposeHeadRoute) {
+          warning.emit('FSTDEP007')
```
Comment (MetCoder95) [inline lib/route.js:352]: The error stated on [this comment](https://github.com/fastify/fastify/pull/2954#issuecomment-819051784) was caused by the order of how the sibling `HEAD` route was being registered for each `GET` one.

As each `GET` route was creating the sibling `HEAD` one before registering a handler for the `after` hook, the `HEAD` route was registering first the hook handler, changing the order on how the validations should be done.

I moved the register of the sibling route on the `after` route hook. Not sure if is the best or should do it after registering the hook handler and not within. Looking forward to having feedback 🙂

File: lib/route.js (hunk 2 above)
Comment (jsumners) [reply lib/route.js:183]: Can we remove this now?

File: lib/route.js (hunk 4 above)
Comment (jsumners) [reply lib/route.js:213]: @mcollina is your concern addressed?

File: test/404s.test.js [hunk 15]
```diff
@@ -830,21 +830,21 @@ test('setNotFoundHandler should not suppress duplicated routes checking', t => {
...
 
-  fastify.listen(0, err => {
-    t.ok(err)
-  })
+    t.fail('setNotFoundHandler should not interfer duplicated route error')
```
Comment (jsumners) [inline test/404s.test.js:844]: "interfer" is not a word I recognize.

Comment (MetCoder95) [reply test/404s.test.js:844]: [This one](https://github.com/fastify/fastify/pull/2954#discussion_r614385800) and here solved in 5bfeeb 👍

File: test/schema-feature.test.js [hunk 16]
```diff
@@ -170,7 +170,7 @@ test('Cannot add schema for query and querystring', t => {
 
 test('Should throw of the schema does not exists in input', t => {
   t.plan(2)
-  const fastify = Fastify()
+  const fastify = Fastify({ exposeHeadRoutes: true })
```
Comment (mcollina) [inline test/schema-feature.test.js:173]: it is `exposeHeadRoutes` is `true` by default.

```suggestion
  const fastify = Fastify()
```

File: lib/route.js (hunk 14 above)
Comment (mcollina) [reply lib/route.js:325]: Good for me

File: test/404s.test.js (hunk 15 above)
Comment (jsumners) [reply test/404s.test.js:844]: I'm still seeing this on GitHub.

File: lib/route.js [hunk 17]
```diff
@@ -208,117 +201,132 @@ function buildRouting (options) {
...
-        return
+      if (opts.version) {
+        warning.emit('FSTDEP006')
+        constraints.version = opts.version
       }
 
-      const { exposeHeadRoute } = opts
-      const hasRouteExposeHeadRouteFlag = exposeHeadRoute != null
-      const shouldExposeHead = hasRouteExposeHeadRouteFlag ? exposeHeadRoute : globalExposeHeadRoutes
+      const headRouteExists = opts.method === 'HEAD' && router.find('HEAD', url, constraints) != null
```
Comment (RafaelGSS) [inline lib/route.js:230]: Not sure why this behavior changed. It's needed for this feature?

File: lib/route.js (hunk 6 above)
Comment (RafaelGSS) [reply lib/route.js:242]: It happens in `next` without this PR?

File: lib/route.js [hunk 18]
```diff
@@ -208,117 +201,132 @@ function buildRouting (options) {
...
+          └── / (GET)
+              / (HEAD)
+              └── / (GET)
+      **/
+
+      if (!headRouteExists) {
+        try {
+          router.on(opts.method, opts.url, { constraints }, routeHandler, context)
+        } catch (error) {
+          const isDuplicatedRoute = Array.isArray(error.message.match(`Method '${opts.method}' already declared for route '${opts.url}'`))
```
Comment (RafaelGSS) [inline lib/route.js:248]: the error instance is not proto of `FST_ERR_DUPLICATED_ROUTE` so you can just check `if (error instanceof FST_ERR_DUPLICATED_ROUTE)`?

File: lib/route.js [hunk 19]
```diff
@@ -208,117 +201,132 @@ function buildRouting (options) {
...
-
-              // Track hooks deprecation markers
-              if (hook === 'preParsing') {
-                // Check for deprecation syntax
-                if (h.length === (h.constructor.name === 'AsyncFunction' ? 2 : 3)) {
-                  warning.emit('FSTDEP004')
-                  bound[kHooksDeprecatedPreParsing] = true
+      this.after((notHandledErr, done) => {
+        // Send context async
+        context.errorHandler = opts.errorHandler || this[kErrorHandler]
```
Comment (RafaelGSS) [inline lib/route.js:259]: Why set the context values here instead of `constructor`? [here](https://github.com/fastify/fastify/pull/2954/files?diff=unified&w=1#diff-cf6996cafd1df0bc168a5c5f3a7470f11fc3aea9f908091d5dd6be2e7b30a2c3L240)

**Review by RafaelGSS — COMMENTED** [review]
Can you add a new test to not call `onRoute` twice when duplicate routes?

File: lib/route.js (hunk 18 above)
Comment (RafaelGSS) [reply lib/route.js:248]: perhaps it's thrown directly by `find-my-way` so you can just ignore it.

File: lib/route.js (hunk 17 above)
Comment (MetCoder95) [reply lib/route.js:230]: When moving to sync registering of routes, yes. This due to the fact that before the sibling `HEAD` route was being added on the avvio's `after` hook, causing no issues as because of the async nature. For instance, wasn't interfering with the proper validations of original `GET` route, getting the validations errors in the right order

File: lib/route.js (hunk 18 above)
Comment (MetCoder95) [reply lib/route.js:248]: Yes, we're relying on `find-my-way`  throwing an error and just decorating it on top. I think the way I'm doing the validation to see if the error message matches with the error we're expecting might not be the best, I'm open to suggestions 🙂 

Regarding ignoring `find-my-way` errors, I'm not so sure. I think is possible that there are validation issues thrown by `find-my-way` and we shouldn't silence it and be handled by the implementation. What do you think?

File: lib/route.js (hunk 19 above)
Comment (MetCoder95) [reply lib/route.js:259]: To be honest, I'm not 100% sure of the reason behind it. Maybe @mcollina knows more about the reason behind it as is based on [his suggestion](https://github.com/fastify/fastify/issues/2842#issuecomment-800962405).

Is also possible that I misunderstood the suggestion, if thats case, I'm open to change the implementation 🙂

File: lib/route.js (hunk 6 above)
Comment (MetCoder95) [reply lib/route.js:245]: Incredibly I was not able to reproduce it on `next` after a rebase. I'll try to figure out if stills reproducible or maybe was part of my changes before getting it stable

File: test/404s.test.js (hunk 15 above)
Comment (MetCoder95) [reply test/404s.test.js]: I made another typo, sorry 😓 
Done in 0659eb5

File: test/schema-feature.test.js (hunk 16 above)
Comment (MetCoder95) [reply test/schema-feature.test.js]: Done in 0659eb5

**Comment by MetCoder95** [comment]
> Can you add a new test to not call `onRoute` twice when duplicate routes?

Shall be covered in that case? I'm wondering because is possible that `onRoute` modify the options of the route being set that's why it runs before setting the route and for instance validating that the route is not a duplicated one. Shall we change the order of it?

**Comment by MetCoder95** [comment]
The TS checks are failing due to [this error](https://github.com/fastify/fastify/pull/2954/checks?check_run_id=2404535065#step:6:11001). It's complaining, I see that also `next` branch is failing because of that, is this expected?

File: lib/route.js (hunk 18 above)
Comment (RafaelGSS) [reply lib/route.js]: What I mean with _"ignore it"_ was ignore my comment if it's throw by `find-my-way`, I'm ok with the currently approach so far :smile:.

**Comment by RafaelGSS** [comment]
> The TS checks are failing due to [this error](https://github.com/fastify/fastify/pull/2954/checks?check_run_id=2404535065#step:6:11001). It's complaining, I see that also `next` branch is failing because of that, is this expected?

It was fixed by:  https://github.com/fastify/fastify/pull/3018

**Comment by RafaelGSS** [comment]
> > Can you add a new test to not call `onRoute` twice when duplicate routes?
> 
> Shall be covered in that case? I'm wondering because is possible that `onRoute` modify the options of the route being set that's why it runs before setting the route and for instance validating that the route is not a duplicated one. Shall we change the order of it?

I see now, I was thinking that the following snipper would call `onRoute` once:

```js
fastify.addHook('onRoute', () => {
  console.log('OnRoute Called')
})

fastify
  .get('/', function (req, reply) {
    reply
      .send({ hello: 'world' })
  })
fastify
  .get('/', function (req, reply) {
    reply
      .send({ hello: 'world' })
  })
```

But, it gives:

```sh
OnRoute Called
OnRoute Called

/fastify/lib/route.js:254
            throw new FST_ERR_DUPLICATED_ROUTE(opts.method, opts.url)
            ^
FastifyError [Error]: Method 'GET' already declared for route '/'
....
```

Personally, I'm ok with the current approach, I don't see any cases that call `onRoute` twice on a duplicate route would be bad or dangerous. However, the current documentation says:

> ### OnRoute
> Triggered when a **new route is registered** [here](https://github.com/fastify/fastify/blob/main/docs/Hooks.md#onroute)

Perhaps, I'd just change to _"Triggered **before** a new route is registered"_ just to avoid confusion. What do you all think of that?

**Comment by mcollina** [comment]
can you merge/rebase on top of next? I've merged master in next.

**Comment by MetCoder95** [comment]
> Personally, I'm ok with the current approach, I don't see any cases that call onRoute twice on a duplicate route would be bad or dangerous. However, the current documentation says:
>
> >OnRoute
> >Triggered when a new route is registered here
>
>Perhaps, I'd just change to "Triggered before a new route is registered" just to avoid confusion. What do you all think of that?

Sure, sounds good to me, has more sense as is indeed triggered before registering the route rather than after it. Let me update the documentation about it 👍 

> can you merge/rebase on top of next? I've merged master in next.
Done, is passing now!

**Comment by MetCoder95** [comment]
I added the changes for the **documentation** and remove the comments on the code, feel free to take another look and share feedback 🙂

**Review by mcollina — APPROVED** [review]
lgtm

**Review by RafaelGSS — APPROVED** [review]
LGTM.

**Comment by mcollina** [comment]
Any more feedbacks on this? @delvedor @jsumners @fastify/core?

**Review by jsumners — APPROVED** [review]
LGTM
